refactor: introduce `BindingResult` to pass errors from rust to js

hyf0 opened the pull request:
Issues of previous error mechanism:

- `BindingOutputs#errors` always got consumed in the js side. It adds unnecessary lazy overhead.
- `BindingOutputs#errors` doesn't force develpoers to handle errors.
- `BindingOutputs#errors` get handled repeatly due to each function doesn't know about whether its caller handles the error or not.

---

This PR introduces `BindingResult` as a unify way, not just for passing `BindingOutput`​, to handle errors passed from rust to js.

- Errors are forced to handle if you want to use the value
- Errors are designed to only need to be handled once.

Related might still need to polished, but I want to set a correct direction first.

hyf0 commented:
* **#6390** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6390?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6390?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#6389** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6389?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6388** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6388?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6387** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6387?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6390).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

IWANABETHATGUY commented on crates/rolldown_binding/src/types/error/mod.rs:
```diff
@@ -1,9 +1,24 @@
+  }
+}
+
+pub type BindingErrorsOr<T> = Either<BindingErrors, T>;
```
What about renaming it to `BindingResult`? It looks like just a `Result` enum in conventional Rust.

hyf0 replied:
Yes. Done.

hyf0 committed:
refactor: introduce `BindingResult` to pass errors from rust to js (#6390)

Issues of previous error mechanism:

- `BindingOutputs#errors` always got consumed in the js side. It adds unnecessary lazy overhead.
- `BindingOutputs#errors` doesn't force develpoers to handle errors.
- `BindingOutputs#errors` get handled repeatly due to each function doesn't know about whether its caller handles the error or not.

---

This PR introduces `BindingResult` as a unify way, not just for passing `BindingOutput`​, to handle errors passed from rust to js.

- Errors are forced to handle if you want to use the value
- Errors are designed to only need to be handled once.

Related might still need to polished, but I want to set a correct direction first.
